### PR TITLE
adds location layer options class

### PR DESCRIPTION
**Pull request by cammace**
Closes https://github.com/mapbox/mapbox-plugins-android/issues/183

Allows you to set style dynamically using the `LocationLayerOptions` class. Also adds all the necessary attributes that were settable in the now deprecated `MyLocationViewSettings`.

### CODE REVIEW

File: plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/StaleStateRunnable.java [hunk 1]
```diff
@@ -27,10 +29,18 @@ private StaleStateRunnable() {
...
+  void addOnLocationStaleListener(@NonNull OnLocationStaleListener onLocationStaleListener) {
     onLocationStaleListeners.add(onLocationStaleListener);
   }
 
+  void removeOnLocationStaleListener(@Nullable OnLocationStaleListener onLocationStaleListener) {
+    if (onLocationStaleListener == null) {
+      onLocationStaleListeners.clear();
+      return;
+    }
+    onLocationStaleListeners.remove(onLocationStaleListener);
```
Comment (tobrun) [inline plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/StaleStateRunnable.java]: this method doesn't do what is expected from its name when a null is passed.

File: plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerPlugin.java [hunk 2]
```diff
@@ -431,11 +439,13 @@ private void disableLocationLayerPlugin() {
    * position.
    */
   @SuppressWarnings( {"MissingPermission"})
-  private void setLastLocation() {
+  private void setLastLocation(int locationLayerMode) {
```
Comment (tobrun) [inline plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerPlugin.java]: why does the end developer need to pass in the mode? isn't this part of the state managed by the plugin itself?

File: plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerPlugin.java [hunk 3]
```diff
@@ -260,6 +266,7 @@ public LocationEngine getLocationEngine() {
    */
   @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
   public void onStop() {
+    StaleStateRunnable.getInstance().removeOnLocationStaleListener(null);
     StaleStateRunnable.getInstance().onStop();
```
Comment (tobrun) [inline plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerPlugin.java]: Calling getInstance twice while one time is sufficient, either assign result/reuse or refactor StateStateRunnable to encapsulate behaviour into one method. If `removeOnLocationStaleListener ` is only used in this code path the latter suggestion would make most sense.

File: plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerOptions.java [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+import android.support.annotation.DrawableRes;
+import android.support.annotation.FloatRange;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.annotation.StyleRes;
+
+import com.google.auto.value.AutoValue;
+
+@AutoValue
+public abstract class LocationLayerOptions implements Parcelable {
```
Comment (tobrun) [inline plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerOptions.java]: missing javadoc on public API from this class.

File: plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/Utils.java [hunk 5]
```diff
@@ -37,20 +44,40 @@ static Bitmap getBitmapFromDrawable(Drawable drawable) {
...
+      Bitmap bitmap = Bitmap.createBitmap(drawable.getIntrinsicWidth(),
+        drawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
       Canvas canvas = new Canvas(bitmap);
       drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
       drawable.draw(canvas);
       return bitmap;
     }
   }
 
+  static Bitmap setShadow(Drawable drawable, float elevation) {
```
Comment (tobrun) [inline plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/Utils.java]: setShadow is not a setter

File: app/src/androidTest/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerTest.java [hunk 6]
```diff
@@ -153,6 +164,38 @@ public void onLocationLayerAction(LocationLayerPlugin locationLayerPlugin, Mapbo
...
+
+  @Test
+  public void whenDrawableChanged_continuesUsingStaleIcons() throws Exception {
+
+  }
+
+  @Test
+  public void whenModeChanged_remainsUsingStaleIcons() throws Exception {
+
+  }
```
Comment (tobrun) [inline app/src/androidTest/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerTest.java]: empty tests

File: app/src/androidTest/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerTest.java [hunk 7]
```diff
@@ -44,6 +50,8 @@
   private MapboxMap mapboxMap;
   private LocationLayerPlugin locationLayerPlugin;
 
+  Location location;
```
Comment (tobrun) [inline app/src/androidTest/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerTest.java]: make private?

File: plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerPlugin.java (hunk 2 above)
Comment (cammace) [reply plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerPlugin.java]: This method isn't exposed to the end dev, it's purpose is to attempt at getting a location value immediately when a mode is set.

**Comment by cammace** [comment]
Thanks for the notes @tobrun, I've fixed up the issues and am ready for a second pass. I also forgot to mention that this PR branched from #264, it might be easier if I merge this into that branch for the final review. Let me know what works for you.

File: plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerPlugin.java (hunk 2 above)
Comment (tobrun) [reply plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerPlugin.java]: That intent is not clear from the method name or the parameter, I would advice coming up with a better name for those.

Comment (tobrun) [reply plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerPlugin.java]: Thanks for clarifying, I'm still a bit confused why you aren't moving this [line](https://github.com/mapbox/mapbox-plugins-android/blob/ae4dc1b3c7f6fba46871ab88f3a5340065c19b95/plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerPlugin.java#L179) to the top of the method and just checking against `this.locationLayerMode` instead. Would this break any business logic?

Comment (cammace) [reply plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerPlugin.java]: Good point, initially testing and not seeing any issues with this approach.

File: plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerPlugin.java [hunk 8]
```diff
@@ -152,12 +152,13 @@ private void initialize() {
    */
   @RequiresPermission(anyOf = {ACCESS_FINE_LOCATION, ACCESS_COARSE_LOCATION})
   public void setLocationLayerEnabled(@LocationLayerMode.Mode int locationLayerMode) {
+    this.locationLayerMode = locationLayerMode;
```
Comment (tobrun) [inline plugin-locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayerPlugin.java]: this change allows to also remove the param in `toggleCameraListener`?

**Pull request merged**